Support differentiation through jax.lax.all_to_all

jpuigcerver opened the pull request:
Credit to @levskaya for the solution.

jpuigcerver committed:
Support differentiation through jax.lax.all_to_all

Credit to @levskaya for the solution.

jekbradbury commented:
This looks right, though maybe we want to add it to the tests somewhere?

levskaya commented:
Q: do we even have an easy way to meaningfully test this now?  is all_to_all implemented on xla:cpu?

hawkinsp commented:
> Q: do we even have an easy way to meaningfully test this now? is all_to_all implemented on xla:cpu?

You can however write tests for these and disable everywhere but on TPU (see `jtu.skip_on_devices`), which we do test internally; I would do that in this case. If it doesn't have a test it will inevitably break and we've had some bugs with wrong derivatives for collectives before.

jekbradbury commented:
I believe CPU supports all-to-all now as of https://github.com/tensorflow/tensorflow/commit/052263c1305c476d1faf470dc0bfdfc24f63d501

jpuigcerver commented:
I'm happy to add some tests. Where should I add those? 
- tests/api_test.py
- tests/lax_test.py
- tests/lax_autodiff_test.py

jpuigcerver committed:
Test gradient of all_to_all

We are testing all_to_all through pswapaxes, since general all_to_all is problematic according to https://github.com/google/jax/issues/1332.

jpuigcerver commented:
Added test in tests/pmap_test.py

jpuigcerver committed:
Removed trailing spaces

jekbradbury commented:
Thanks, Anselm and Joan!
